src/librustc_privacy/lib.rs

```diff
@@ -1092,58 +1076,103 @@ impl<'a, 'tcx, 'v> Visitor<'v> for PrivateItemsInPublicInterfacesVisitor<'a, 'tc
             hir::ItemUse(..) | hir::ItemUseGlob(..) => {}
             // Subitems of these items have inherited publicity
             hir::ItemConst(..) | hir::ItemStatic(..) | hir::ItemFn(..) |
-            hir::ItemEnum(..) | hir::ItemTrait(..) | hir::ItemTy(..) => {
-                check.required_visibility = item_visibility;
-                check.visit_item(item);
+            hir::ItemTy(..) => {
+                self.check(item.id, item_visibility).generics().predicates().item_type();
+
+                // Take `impl Trait` into account.
```

Comment:
I asked question 2 in https://github.com/rust-lang/rust/pull/37676#issuecomment-260162743 because this comment was confusing.
I could never guess that `impl Trait` has its own item predicates.